packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts

```diff
@@ -83,6 +89,56 @@ export class FederatedQuadSource implements ILazyQuadSource {
     return term;
   }
 
+  /**
+   * If the given term is a blank node, return a deterministic named node for it
+   * based on the source id and the blank node value.
+   * @param term Any RDF term.
+   * @param sourceId A numerical source identifier.
+   * @return If the given term was a blank node, this will return a skolemized named node, otherwise the original term.
+   */
+  public static skolemizeTerm(term: RDF.Term, sourceId: number): RDF.Term | BlankNodeSkolemizable {
+    if (term.termType === 'BlankNode') {
+      const skolemized = `urn:comunica_skolem:source_${sourceId}:${term.value}`;
+      return new BlankNodeSkolemizable(skolemized, DataFactory.namedNode(skolemized));
+    }
+    return term;
+  }
+
+  /**
+   * Skolemize all terms in the given quad.
+   * @param quad An RDF quad.
+   * @param sourceId A numerical source identifier.
+   * @return The skolemized quad.
+   */
+  public static skolemizeQuad<Q extends BaseQuad = Quad>(quad: Q, sourceId: number): Q {
+    return mapTerms(quad, (term) => FederatedQuadSource.skolemizeTerm(term, sourceId));
+  }
+
+  /**
+   * If a given term was a skolemized named node for the given source id,
+   * deskolemize it again to a blank node.
+   * If the given term was a skolemized named node for another source, return false.
+   * If the given term was not a skolemized named node, return the original term.
+   * @param term Any RDF term.
+   * @param sourceId A numerical source identifier.
+   */
+  public static deskolemizeTerm(term: RDF.Term, sourceId: number): RDF.Term | false {
+    if (term.termType === 'NamedNode' || term.termType === 'BlankNode') {
+      const match = /^urn:comunica_skolem:source_([0-9]+):(.+)$/.exec(term.value);
+      if (match) {
+        // We had a skolemized term
+        if (parseInt(match[1], 10) === sourceId) {
```

Comment:
We can skip the parsing if we give sources `string` IDs (which _could_ be stringified numbers, all fine).